Phase 2 blog post (#3562)
jenkins-infra/jenkins.io: content/blog/2020/07/2020-07-27-machine-learning-plugin-coding-phase2.adoc

---
layout: post
title: "Machine Learning Plugin project - Coding Phase 2 blog post"
tags:
- machinelearning
- datascience
- plugins
- communitybonding
- gsoc
- gsoc2020
author: loghijiaha

> description: "Coding Phase 2 of Google Summer of Code 2020"

Reviewer: ```suggestion
description: "Google Summer of Code 2020 - Machine Learning Plugin project update for Coding Phase 2"
```

Would be also great to add a sentence summarizing the key updates

opengraph:

>   image: /images/gsoc/opengraph.png

Reviewer: A custom social media image would be awesome, but it is not blocking this PR

---

image::/images/gsoc/jenkins-gsoc-logo_small.png[align="center"]

Welcome back folks!

This blog post is about my coding phase 2 in Jenkins link:/projects/gsoc/2020/projects/machine-learning/[Machine Learning Plugin] for this GSoC 2020.
After successfully passing the evaluation and demo in the phase 1, our team went ahead for facing the challenges in phase 2.

== Summary

This phase of coding was well spent by documentation and by fixing many bugs.
As the main feature of connecting to an IPython Kernel is done in phase 1, we were able to focus on fixing minor/major bugs and documenting for the users.
According to the link:https://issues.jenkins-ci.org/browse/JENKINS-62927[JENKINS-62927] issue, a Docker agent was built to facilitate users without concerning plugin dependencies in python.
In the act of deprecation of Python 2, we ported our plugin to support Python 3.

> We have tested our plugin in conda, venv and windows environments.

Reviewer: ```suggestion
We have tested our plugin in Conda, venv and Windows environments.
```

Machine learning plugin has successfully passed the end to end test. A feature for a code editor is needed for further discussion/analysis as we have done a simple editor that may be useful in other ways in the future. link:https://github.com/jenkinsci/machine-learning-plugin/pull/35[PR#35]

== Main features of Machine Learning plugin

- Run Jupyter notebook, (Zeppelin) JSON and Python files
- Run Python code directly
- Convert Jupyter Notebooks to Python and JSON
- Configure IPython kernel properties
- Support to execute Notebooks/Python on Agent
- Support for Windows and Linux

== Upcoming features

- Extract graph/map/images from the code
- Save artifacts according to the step name
- Generate reports for corresponding build

== Future improvements

> 

Reviewer: Do you plan to do it during GSoC? If not, maybe it makes sense to say it explicitly to highlight there is no committment.

- Usage of JupyterRestClient
- Support for multiple language kernels

=== Docker agent

> The following dockerfile can be used to build the docker container as an agent for machine learning plugin. This docker agent can be used to run  notebooks or python scripts.

Reviewer: ```suggestion
The following Dockerfile can be used to build the Docker container as an agent for the Machine Learning plugin. This docker agent can be used to run  notebooks or python scripts.
```

---
layout: post
title: "Machine Learning Plugin project - Coding Phase 2 blog post"
tags:
- machinelearning
- datascience
- plugins
- communitybonding
- gsoc
- gsoc2020
author: loghijiaha

> description: "Coding Phase 2 of Google Summer of Code  2020"

Reviewer: ```suggestion
description: "Coding Phase 2 of Google Summer of Code 2020"
```

Would be great to add a short summary here (could be used as a sample for promotion in social media)

opengraph:
  image: /images/gsoc/opengraph.png
---

image::/images/gsoc/jenkins-gsoc-logo_small.png[align="center"]

> Welcome back forks!

Reviewer: ```suggestion
Welcome back folks!
```

Just in case it was a typo :)

Reviewer 2: I also wasn't sure if it was intentional or not :)

This blog post is about my coding phase 2 in Jenkins link:/projects/gsoc/2020/projects/machine-learning/[Machine Learning Plugin] for this GSoC 2020.

> 

Reviewer: ```suggestion
```

No need to start a new paragraph IMHO

=== Documentation and End to End testing

A well explained documentation was published in the repository. A guided tutorial to run a notebook checked out from a git repo in an agent was included in the link:https://github.com/jenkinsci/machine-learning-plugin/blob/master/docs/ML-docker-agent.adoc[docs] page. Mentors helped to test our plugin in both Linux and Windows.

=== Code editor with rebuild feature

Code editor was filtered as a nice to have feature in the design document. After grabbing the idea of Jenkinsfile replay editor, I could do the same for the code. At the same time, when we are getting the source code from git, it is not an elegant way of editing code in the original code. After the discussion, we had to leave the PR open that may have use cases in the future if needed.

=== Jenkins LTS update

The plugin has been updated to support Jenkins LTS 2.204.1 as 2.164.3 had some problems with installing pipeline supported API/plugin

== Installation for experimental version

> . Enable the link:/doc/developer/publishing/releasing-experimental-updates/#using-the-experimental-update-center[experimental update center]

Reviewer: Maybe it makes sense to cut a release candidate (e.g. version `1.0-rc-1`) so that the plugin appears in the main update centers and on the plugin site. It is not required for the blogpost, but something for a discussion @Loghijiaha @kinow

Reviewer 2: If that's necessary I'd be fine with that. Only concern I have is if this gets more attention now, @Loghijiaha could have to spend more time helping/answering user questions (which already happened twice so far I think, even though it's in the experimental). Maybe when we get closer to the end of the GSoC?

There are also updates that are breaking backward compatibility, so `rc-1` and `rc-2` would not be fully compatible (which I think some users would expect, but we could also make it explicit in the plugin readme or description that appears in the update center)

Reviewer: > If that's necessary I'd be fine with that. 

It is not necessary, up to the project team to decide

. Search for `Machine Learning Plugin` and check the box along it.
. Click on `Install without restart`

The plugin should now be installed on your system.

> == Knowledge transfer

Reviewer: IMHO it rather belongs to the plugin's documentation, but up to you

---
layout: post
title: "Machine Learning Plugin project - Coding Phase 2 blog post"
tags:
- machinelearning
- datascience
- plugins
- communitybonding
- gsoc
- gsoc2020
author: loghijiaha
description: "Coding Phase 2 of Google Summer of Code  2020"
opengraph:
  image: /images/gsoc/opengraph.png
---

image::/images/gsoc/jenkins-gsoc-logo_small.png[align="center"]

Welcome back forks!

> This blog post is briefing my coding phase 2 in Jenkins link:/projects/gsoc/2020/projects/machine-learning/[Machine Learning Plugin] for this GSoC 2020.

Reviewer 2: Maybe "this blog post is about my", or "this brief blog post is about my"? We use briefing in Portuguese too (rarely) but normally in military context and as noun, I think that's the same in English.

> After successfully passed the evaluation and demo in the phase 1, our team were ahead for facing the challenges in phase 2.

Reviewer 2: I think "After successfully passing", or "After having successfully passed"...

And "our team went ahead".

> === Summary

Reviewer 3: ```suggestion
== Summary
```

> This phase of coding was well spent with fixing lots of bugs and documentation. As the main feature of connecting to a IPython Kernel is done in the phase 1, we were able to focus on fixing minor/major bugs and documenting for the users. According to the JENKINS-62927 issue, a Docker agent was built to facilitate users without concerning plugin dependencies in python. In the act of deprecation of Python 2, we were indeed to port our plugin to support Python 3. We have tested our plugin in conda, venv and windows environments.

Reviewer 3: Sentence per line is easier to code review, but not strictly required.  Phrasing changes are all optional.
```suggestion
This phase of coding was well spent by documentation and by fixing many bugs.
As the main feature of connecting to an IPython Kernel is done in phase 1, we were able to focus on fixing minor/major bugs and documenting for the users.
According to the link:https://issues.jenkins-ci.org/browse/JENKINS-62927[JENKINS-62927] issue, a Docker agent was built to facilitate users without concerning plugin dependencies in python.
In the act of deprecation of Python 2, we ported our plugin to support Python 3.
We have tested our plugin in conda, venv and windows environments.
```

Machine learning plugin has successfully passed the end to end test. A feature for a code editor is needed for further discussion/analysis as we have done a simple editor that may be useful in other ways in the future. link:https://github.com/jenkinsci/machine-learning-plugin/pull/35[PR#35]

> ==== Main features of Machine Learning plugin

Reviewer 3: ```suggestion
== Main features of Machine Learning plugin
```

- Run Jupyter notebook, (Zeppelin) JSON and Python files
- Run Python code directly
- Convert Jupyter Notebooks to Python and JSON
- Configure IPython kernel properties
- Support to execute Notebooks/Python on Agent
- Support for Windows and Linux

> ====== Upcoming features

Reviewer 3: ```suggestion
== Upcoming features
```

- Extract graph/map/images from the code
- Save artifacts according to the step name
- Generate reports for corresponding build

> ====== Future improvements

Reviewer 3: ```suggestion
== Future improvements
```

- Usage of JupyterRestClient
- Support for multiple language kernels

> ==== Docker agent

Reviewer 3: ```suggestion
== Docker agent
```

The following dockerfile can be used to build the docker container as an agent for machine learning plugin. This docker agent can be used to run  notebooks or python scripts.

.Dockerfile

```
FROM jenkins/agent:latest

MAINTAINER Loghi <loghijiaha@gmail.com>

USER root

RUN apt update && apt install --no-install-recommends python3 -y \
    python3-pip \
    && rm -rf /var/lib/apt/lists/*

COPY requirements.txt /requirements.txt

RUN pip3 install --upgrade pip setuptools && \
    pip3 install --no-cache-dir -r /requirements.txt && \
    ln -sf /usr/bin/python3 /usr/bin/python && \
    ln -sf /usr/bin/pip3 /usr/bin/pip

USER jenkins
```

> ==== Ported to Python 3

Reviewer 3: Is this considered a subsection of the "Docker agent" section or is it an independent section in the blog post?
```suggestion
=== Ported to Python 3
```

As discussed in the previous meeting, we concluded that the plugin should support Python 3 as Python 2.7+ has been deprecated since the beginning of 2020. Pull request for docker agent should be also ported to Python 3 support.

> ==== Jupyter Rest Client API

Reviewer 3: ```suggestion
=== Jupyter Rest Client API
```

The Jupyter Notebook server API seemed to be promising that it can be also used to run notebooks and codes. There were 3 api implementations that were merged in the master. But we had to focus on what was proposed in the design document and had to finish all must-have issues/works. Jupyter REST client was left for future implementation. It is also a good start to contribute to the plugin from the community.

> ==== Fixed bugs for running in agent

Reviewer 3: ```suggestion
=== Fixed bugs for running in agent
```

> There were a few bugs related to the file path of notebooks while building a job. The most problem was caused by the python dependencies needed to connect to a IPython kernel. All issues/bugs were fixed before the timeline given.

Reviewer 3: ```suggestion
There were a few bugs related to the file path of notebooks while building a job. Most problems were caused by the python dependencies needed to connect to a IPython kernel. All issues/bugs were fixed before the timeline given.
```

> There were a few bugs related to the file path of notebooks while building a job. The most problem was caused by the python dependencies needed to connect to a IPython kernel. All issues/bugs were fixed before the timeline given.

Reviewer 2: Instead of "The most problem" I think "The main problem", or "The major problem" perhaps?

Reviewer 2: (And well done on keeping ahead of the schedule @Loghijiaha !)

> ==== R support as a future improvement

Reviewer 3: ```suggestion
=== R support as a future improvement
```

> This is what we tried to give a glimpse of knowledge that this plugin can be extended for multi language support in the future. There was a conclusion that the kernel should be selected dynamically using the mime type of notebook, instead of scripting the same code for each kernel.

Reviewer 2: The mime type of a script I think (e.g. `train_model.r`, or `train_model.groovy`). We could identify the language, and use the right kernel to execute it (ideally leaving that up to Zeppelin and IPython as we discussed, instead of adding code for each language to the plugin).

For notebooks, we probably will need to parse/read the [magic commands](https://ipython.readthedocs.io/en/stable/interactive/magics.html), like `%%javascript`. And we will also need to consider the case when the notebook is completely written for JavaScript or for another language. In that case, the user would have chosen to use the JavaScript Kernel, and written in the cells without the magic line (that's how [I do to try JS features/libs](https://github.com/kinow/notebooks/blob/master/javascript/enumify/Enumify.ipynb)).

Reviewer 2: Click on "Raw" on GitHub, and search for `kernelspec` and `language_info`, that's where that notebook is defining that it should run with JavaScript. That way I can have packages installed via `package.json`/npm.

Author: @kinow our plugin is capable of running magic commands as well. But we should consider running the train_model.r or train_model.groovy etc type of files. If we have time, we will plan for multi-language support in our plugin.

> ==== Documentation and End to End testing

Reviewer 3: ```suggestion
=== Documentation and End to End testing
```

A well explained documentation was published in the repository. A guided tutorial to run a notebook checked out from a git repo in an agent was included in the link:https://github.com/jenkinsci/machine-learning-plugin/blob/master/docs/ML-docker-agent.adoc[docs] page. Mentors helped to test our plugin in both Linux and Windows.

> ==== Code editor with rebuild feature

Reviewer 3: ```suggestion
=== Code editor with rebuild feature
```

Code editor was filtered as a nice to have feature in the design document. After grabbing the idea of Jenkinsfile replay editor, I could do the same for the code. At the same time, when we are getting the source code from git, it is not an elegant way of editing code in the original code. After the discussion, we had to leave the PR open that may have use cases in the future if needed.

> ==== Jenkins LTS update

Reviewer 3: ```suggestion
=== Jenkins LTS update
```

The plugin has been updated to support Jenkins LTS 2.204.1 as 2.164.3 had some problems with installing pipeline supported API/plugin

> === Installation for experimental version

Reviewer 3: I believe this is a top level heading while the preceding headings were subtopics of "Results" or "Future Plans"
```suggestion
== Installation for experimental version
```

> Follow the steps in Jenkins to download and install the plugin:
> 
> . Select `Manage Jenkins`
> 
> . Select `Manage Plugins`
> 
> . Go to `Advanced` tab
> 
> . Configure the Update Site URL as: https://updates.jenkins.io/experimental/update-center.json
> 
> . Click on `Submit`, and then press the `Check Now` button.
> 
> . Go to `Available` tab.
> 
> . Search for `Machine Learning Plugin` and check the box along it.
> 
> . Click on `Install without restart`
> 
> The plugin should now be installed on your system.

Reviewer 3: You might consider simplifying this with
```suggestion
. Enable the link:/doc/developer/publishing/releasing-experimental-updates/#using-the-experimental-update-center[experimental update center]
. Search for `Machine Learning Plugin` and check the box along it.
. Click on `Install without restart`

The plugin should now be installed on your system.
```

> === Knowledge transfer

Reviewer 3: ```suggestion
== Knowledge transfer
```

> ==== How to debug the code through IntelliJ

Reviewer 3: ```suggestion
=== How to debug the code through IntelliJ
```

- Edit configuration → Add new Configuration → Maven
- Command line → type `hpi:run`
- Click the debug icon on the toolbar or go to `Run` menu then `Debug`

> ==== How to setup to test the plugin

Reviewer 3: ```suggestion
=== How to setup to test the plugin
```

- Setup JDK 8 and Maven 3.5.*
- Create a directory `$ mkdir machine-learning-plugin`
- Create a virtual environment `$ virtualenv venv`
- Activate your virtual environment $ source venv/bin/activate
- Run `$ which python` to ensure your python path
- `$ git clone https://github.com/jenkinsci/machine-learning-plugin.git`
- Run `$ mvn clean install` from the machine-learning-plugin directory
- Run `$ mvn hpi:run` to start Jenkins with the plugin
- Set up the builder with localhost and other parameters
- Create a job
- Write python code like print(“plugin works”)
- Build the job

> === Resources

Reviewer 3: ```suggestion
== Resources
```
